### DISCUSSION TITLE
allow .distinct() to specify a column for DISTINCT ON

### DISCUSSION THREAD

**Discussion by williamflaherty** [piccolo-orm/piccolo#808] [Ideas]
As the title says, would be great if you could specify a column in TableName.select().distinct() to handle distinct on or something similar? 

I started poking around to add it myself but got stopped up by order_by needing to be in the same order issues from asyncpg. (I'm far from a PostgreSQL expert).

**Comment by dantownsend** [comment]
This would be a nice feature.

Looking at the [docs](https://www.postgresql.org/docs/current/sql-select.html#SQL-DISTINCT) it seems like the column specified in the `DISTINCT ON` must also be the first column specified in the  `ORDER BY`.

So in this example, `location`:

```sql
SELECT DISTINCT ON (location) location, time, report
    FROM weather_reports
    ORDER BY location, time DESC;
```

This might be why `asyncpg` is complaining.

Feel free to open a PR if you want - doesn't have to be perfect.

**Reply by williamflaherty** [reply]
When I have a chance to put it together I'll make a PR, thanks @dantownsend!

**Reply by sinisaos** [reply]
Sorry to interrupt this discussion. When I read about `DISTINCT ON` online, many people use that Postgres feature and find it useful for aggregation. [Here is my attempt](https://github.com/sinisaos/piccolo/commit/c513b43d6a81d4dcaeed23efcf450a7ee161ac6c) at a Piccolo implementation. If it's not good enough for PR, I wouldn't bother with it anymore and leave it to someone else.

**Reply by dantownsend** [reply]
@sinisaos Thanks for that. 👍 

You added a separate `distinct_on` method, but what about using the existing `distinct` method and adding an optional `on` argument which accepts a column:

```python
# Get the start date of the next concert for each band.
await Concert.select(
    Concert.band,
    Concert.start_date
).distinct(
    on=Concert.band
).order_by(
    Concert.band
).order_by(
    Concert.start_date,
    ascending=False
)
```

We should raise a `ValueError` if trying to use this with any DB except Postgres, as I think it's the only DB to support it.

**Reply by dantownsend** [reply]
The reason I prefer having just the `distinct` method with an `on` argument is it might make the logic a bit easier. If we have both `distinct` and `distinct_on` we can only allow one of them.

**Reply by sinisaos** [reply]
@dantownsend Quite the opposite, I wanted to make as big a distinction as possible between the `distinct` and `distinct_on` methods and that's why I made two methods because `distinct` and `distinct_on` are quite different clauses. In `distinct_on`, the first column in `self.columns_delegate.selected_columns` is always an expression in `DISTINCT_ON` clause.

> If we have both `distinct` and `distinct_on` we can only allow one of them.

We can have only one clause, not both

```python
select = (
    "SELECT DISTINCT" # distinct
    if self.distinct_delegate._distinct
    else "SELECT DISTINCT ON" # distinct_on
    if self.distinct_on_delegate._distinct_on
    else "SELECT" # normal select
)
```
It's somehow more understandable to me with two different methods, but if you thought that the implementation should be different, I'll delete the branch.

**Reply by dantownsend** [reply]
@sinisaos I see why you did it that way, and it's valid solution.

I like to try and mirror the Postgres docs as much as possible, and even though `DISTINCT ON` and `DISTINCT` do quite different things, in the docs `ON` is more of an option for `DISTINCT`, which is why I like the approach of adding an argument to the `distinct` method.

<img width="506" alt="Screenshot 2023-04-24 at 14 48 14" src="https://user-images.githubusercontent.com/350976/234016016-dfd6ada6-7609-4d86-9430-82651057d05e.png">

In your branch you've made it very easy to use by not having to specify the column, but with `DISTINCT ON` you can specify multiple columns, so we need to let the user specify it otherwise we're not exposing all of the functionality.

**Reply by sinisaos** [reply]
> In your branch you've made it very easy to use by not having to specify the column, but with `DISTINCT ON` you can specify multiple columns, so we need to let the user specify it otherwise we're not exposing all of the functionality.

Now you can also specify multiple columns like this

```python
Piccolo:
Band.select(Band.name, Band.popularity, #or any other column#)
.distinct_on().order_by(Band.name) 

SQL:
SELECT DISTINCT ON (name  # this is self.columns_delegate.selected_columns #) 
name, popularity FROM band ORDER BY name;

```
Now you can also specify multiple `order_by` or no `order_by `at all.

@dantownsend No problem. Thanks for explaining what you think about the implementation, so I'll leave it up to you or someone else. Thank you.

**Reply by dantownsend** [reply]
@sinisaos It's just a matter of preference - what you've done it great. But I think making the user explicitly specify everything makes it work for more edge cases, but has the downside of being less convenient. Don't delete what you've done - I'll have a play around with it, and will create a PR based off it with this alternative approach so you can see what I'm suggesting.

**Reply by dantownsend** [reply]
Here's a draft PR: https://github.com/piccolo-orm/piccolo/pull/813

**Comment by dantownsend** [comment]
@williamflaherty We've made some progress on this: https://github.com/piccolo-orm/piccolo/pull/813

I know you already did some thinking about it, so would appreciate your thoughts on the current solution.

I've sent you an invite with the `triage` role, otherwise I can't mention you in the PR. It's a strange limitation of GitHub 🤦
